File: app/src/main/java/pl/szczodrzynski/edziennik/utils/models/Notification.java
package pl.szczodrzynski.edziennik.utils.models;

import android.content.Context;
import android.content.Intent;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonPrimitive;

import java.util.Map;
import java.util.Random;

import pl.szczodrzynski.edziennik.R;

import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_AUTO_ARCHIVING;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_ERROR;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_FEEDBACK_MESSAGE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_GENERAL;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_LUCKY_NUMBER;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_ANNOUNCEMENT;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_ATTENDANCE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_EVENT;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_GRADE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_HOMEWORK;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_MESSAGE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_NOTICE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_NEW_SHARED_EVENT;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_SERVER_MESSAGE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_TIMETABLE_CHANGED;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_TIMETABLE_LESSON_CHANGE;
import static pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.TYPE_UPDATE;

public class Notification {
    public int profileId;
    public String title;
    public boolean notified;
    public boolean seen;
    public int redirectFragmentId;
    public JsonObject extras;
    public int type;
    public String text;
    public long addedDate;
    public int id;

    public Notification(Context context, String text) {
        this.profileId = -1;
        this.title = context.getString(R.string.notification);
        this.type = TYPE_GENERAL;
        this.text = text;
        this.redirectFragmentId = -1;
        this.extras = new JsonObject();
        this.notified = false;
        this.seen = false;
        this.addedDate = System.currentTimeMillis();
        this.id = new Random().nextInt(Integer.MAX_VALUE);
    }

    public Notification withProfileData(int profileId, String profileName) {
        this.profileId = profileId;
        this.title = profileName;
        return this;
    }

    public Notification withProfileData(int profileId) {
        this.profileId = profileId;
        return this;
    }

    public Notification withTitle(String title) {
        this.title = title;
        return this;
    }

    public Notification withType(int type) {
        this.type = type;
        return this;
    }

    public Notification withFragmentRedirect(int redirectFragmentId) {
        this.redirectFragmentId = redirectFragmentId;
        return this;
    }

    public Notification withLongExtra(String key, long value) {
        this.extras.addProperty(key, value);
        return this;
    }

    public Notification withStringExtra(String key, String value) {
        this.extras.addProperty(key, value);
        return this;
    }

    public Notification withAddedDate(long addedDate) {
        this.addedDate = addedDate;
        return this;
    }

    public static String stringType(Context context, int errorCode)
    {
        switch (errorCode) {
            case TYPE_UPDATE:
                return context.getString(R.string.notification_type_update);
            case TYPE_ERROR:
                return context.getString(R.string.notification_type_error);
            case TYPE_TIMETABLE_CHANGED:
                return context.getString(R.string.notification_type_timetable_change);
            case TYPE_TIMETABLE_LESSON_CHANGE:
                return context.getString(R.string.notification_type_timetable_lesson_change);
            case TYPE_NEW_GRADE:
                return context.getString(R.string.notification_type_new_grade);
            case TYPE_NEW_EVENT:
                return context.getString(R.string.notification_type_new_event);
            case TYPE_NEW_HOMEWORK:
                return context.getString(R.string.notification_type_new_homework);
            case TYPE_NEW_SHARED_EVENT:
                return context.getString(R.string.notification_type_new_shared_event);
            case TYPE_NEW_MESSAGE:
                return context.getString(R.string.notification_type_new_message);
            case TYPE_NEW_NOTICE:
                return context.getString(R.string.notification_type_notice);
            case TYPE_NEW_ATTENDANCE:
                return context.getString(R.string.notification_type_attendance);
            case TYPE_SERVER_MESSAGE:
                return context.getString(R.string.notification_type_server_message);
            case TYPE_LUCKY_NUMBER:
                return context.getString(R.string.notification_type_lucky_number);
            case TYPE_FEEDBACK_MESSAGE:
                return context.getString(R.string.notification_type_feedback_message);
            case TYPE_NEW_ANNOUNCEMENT:
                return context.getString(R.string.notification_type_new_announcement);
            case TYPE_AUTO_ARCHIVING:
                return context.getString(R.string.notification_type_auto_archiving);
            default:
            case TYPE_GENERAL:
                return context.getString(R.string.notification_type_general);
        }
    }

    public void fillIntent(Intent intent) {
        if (profileId != -1)
            intent.putExtra("profileId", profileId);

        /*if (redirectFragmentId == DRAWER_ITEM_MESSAGES)
            redirectFragmentId = DRAWER_ITEM_MESSAGES_INBOX;*/

        if (redirectFragmentId != -1)
            intent.putExtra("fragmentId", redirectFragmentId);

        try {
            for (Map.Entry<String, JsonElement> entry: extras.entrySet()) {
                JsonElement value = entry.getValue();
                if (!value.isJsonPrimitive())
                    continue;
                JsonPrimitive primitive = value.getAsJsonPrimitive();
                if (primitive.isNumber()) {
                    intent.putExtra(entry.getKey(), primitive.getAsLong());
                }
                else if (primitive.isString()) {
                    intent.putExtra(entry.getKey(), primitive.getAsString());
                }
            }
        }
        catch (NullPointerException e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/modules/notification/Notification.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2019-10-18.
 */

package pl.szczodrzynski.edziennik.data.db.modules.notification

import android.content.Context
import android.content.Intent
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.gson.JsonObject
import pl.szczodrzynski.edziennik.R
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_AUTO_ARCHIVING
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_ERROR
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_FEEDBACK_MESSAGE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_GENERAL
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_LUCKY_NUMBER
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_ANNOUNCEMENT
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_ATTENDANCE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_EVENT
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_GRADE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_HOMEWORK
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_MESSAGE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_NOTICE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_NEW_SHARED_EVENT
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_SERVER_MESSAGE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_TIMETABLE_CHANGED
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_TIMETABLE_LESSON_CHANGE
import pl.szczodrzynski.edziennik.data.db.modules.notification.Notification.Companion.TYPE_UPDATE

@Entity(tableName = "notifications")
data class Notification(
        @PrimaryKey(autoGenerate = true)
        val id: Int = 0,

        val title: String,
        val text: String,

        val type: Int,

        val profileId: Int?,
        val profileName: String?,

        var posted: Boolean = false,

        var viewId: Int? = null,
        var extras: JsonObject? = null,

        val addedDate: Long = System.currentTimeMillis()
) {
    companion object {
        const val TYPE_GENERAL = 0
        const val TYPE_UPDATE = 1
        const val TYPE_ERROR = 2
        const val TYPE_TIMETABLE_CHANGED = 3
        const val TYPE_TIMETABLE_LESSON_CHANGE = 4
        const val TYPE_NEW_GRADE = 5
        const val TYPE_NEW_EVENT = 6
        const val TYPE_NEW_HOMEWORK = 10
        const val TYPE_NEW_SHARED_EVENT = 7
        const val TYPE_NEW_SHARED_HOMEWORK = 12
        const val TYPE_NEW_MESSAGE = 8
        const val TYPE_NEW_NOTICE = 9
        const val TYPE_NEW_ATTENDANCE = 13
        const val TYPE_SERVER_MESSAGE = 11
        const val TYPE_LUCKY_NUMBER = 14
        const val TYPE_NEW_ANNOUNCEMENT = 15
        const val TYPE_FEEDBACK_MESSAGE = 16
        const val TYPE_AUTO_ARCHIVING = 17
    }

    fun addExtra(key: String, value: Long?): Notification {
        extras = extras ?: JsonObject()
        extras?.addProperty(key, value)
        return this
    }
    fun addExtra(key: String, value: String?): Notification {
        extras = extras ?: JsonObject()
        extras?.addProperty(key, value)
        return this
    }

    fun fillIntent(intent: Intent) {
        if (profileId != -1)
            intent.putExtra("profileId", profileId)
        if (viewId != -1)
            intent.putExtra("fragmentId", viewId)
        try {
            extras?.entrySet()?.forEach { (key, value) ->
                if (!value.isJsonPrimitive)
                    return@forEach
                val primitive = value.asJsonPrimitive
                if (primitive.isNumber) {
                    intent.putExtra(key, primitive.asLong)
                } else if (primitive.isString) {
                    intent.putExtra(key, primitive.asString)
                }
            }
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }
}

fun Context.getNotificationTitle(type: Int): String {
    return getString(when (type) {
        TYPE_UPDATE -> R.string.notification_type_update
        TYPE_ERROR -> R.string.notification_type_error
        TYPE_TIMETABLE_CHANGED -> R.string.notification_type_timetable_change
        TYPE_TIMETABLE_LESSON_CHANGE -> R.string.notification_type_timetable_lesson_change
        TYPE_NEW_GRADE -> R.string.notification_type_new_grade
        TYPE_NEW_EVENT -> R.string.notification_type_new_event
        TYPE_NEW_HOMEWORK -> R.string.notification_type_new_homework
        TYPE_NEW_SHARED_EVENT -> R.string.notification_type_new_shared_event
        TYPE_NEW_MESSAGE -> R.string.notification_type_new_message
        TYPE_NEW_NOTICE -> R.string.notification_type_notice
        TYPE_NEW_ATTENDANCE -> R.string.notification_type_attendance
        TYPE_SERVER_MESSAGE -> R.string.notification_type_server_message
        TYPE_LUCKY_NUMBER -> R.string.notification_type_lucky_number
        TYPE_FEEDBACK_MESSAGE -> R.string.notification_type_feedback_message
        TYPE_NEW_ANNOUNCEMENT -> R.string.notification_type_new_announcement
        TYPE_AUTO_ARCHIVING -> R.string.notification_type_auto_archiving
        TYPE_GENERAL -> R.string.notification_type_general
        else -> R.string.notification_type_general
    })
}